### ISSUE TITLE
Feature Request: Keep sending when a transaction done.

### ISSUE THREAD

**Issue by kehao95** [warner/magic-wormhole#230]

**Comment by warner** [comment]
I *think* this is a duplicate of #76, but could you take a look at the discussion there and see if you agree? (actually reusing the same code multiple times is a security problem, but there might be other ways to improve the send-many or send-one-to-many flow)

**Comment by kehao95** [comment]
Sorry, I was not very clear when I created this issue. But I think this is different from #76.
#76 was talking about easy connection to a known person. So the possible way to solve this is 'reusable code' or 'specify a code'. And you pointed out keep connection may be a better way than using the same code.
And **this issue** is created when I try to figure out how to send from one to many. Actually, my suggestion does not mean using the same code from day to day. Besides, I believe a randomly generated code should only be valid for one 1-to-N transaction. The code is still short lived.

So I can think of two possible ways to improve this:
1. implemented an option like `--reuse-times times`. `times` should be equals to N. You can type the same code on N computers to receive this file. After finishing N transactions the sender should automatically close. Mostly, I believe this workflow can be done within half an hour, supposing the file size and N are relatively small.
2.  implemented an option like `--code-timeout mins`. And sender will allow all connections (restart waiting for connection using the same code, when a transaction is done) within the timeout. And automatically close when timeout.

Both ways above can make sure the code is short-lived when used properly. And you can improve the security by using both limits. For example, set default timeout to 30 mins when using `--reuse-times`. 
The only dangerous scene should happen when a stupid ass deliberately set both `--reuse-times infinite` and `--code-timeout infinite`. On that circumstance, he may want to use the wormhole as a Content Deliver Sevice. I think it's also OK to let him do that.

**Comment by warner** [comment]
Ah, interesting. Let's see, allowing the same code to be reused multiple times also allows an attacker multiple guesses. We could mitigate this by making multiple-use codes longer: if you allow two uses, add one bit of entropy, if you allow 16 uses, add four bits, etc. Three words would be enough to cover 256 uses without improving the attacker's overall probability of success.

You couldn't allow an unbounded number of recipients, even with a timeout, since an attacker could just exhaustively test the entire namespace until they find something that works. So we'd need a finite number of recipients in addition to the time limit. (And the time limit is a good UX thing, since with the single recipient you know when you're done, whereas with multiple recipients you probably won't).

The trickier part is that the "nameplate" (the numeric channel identifier at the beginning of the wormhole code: the "4" in "4-purple-sausages") is used to identify which "mailbox" or queue the client is using. If three or more clients all use the same nameplate, their messages will get jumbled up.

So we'd need some change to the protocol to allow this to work. Maybe the sender doesn't put anything in the mailbox at first, and waits to see a PAKE message from one of the receivers. Then the sender would need to narrowcast a PAKE message for just that one receiver (while continuing to listen for PAKE from other receivers). It'd be nice if these could use separate mailboxes.. I think the server would need to be involved.

Let me think about it.. maybe it's possible.

**Comment by KeithBierman** [comment]
perhaps introducing an interactive mode ... the code is still good only once ... but that is one interactive session ... keep it open until either side closes their connection.

**Comment by meejah** [comment]
> perhaps introducing an interactive mode ... the code is still good only once ... but that is one interactive session ... keep it open until either side closes their connection.

This is one of the use-cases for Dilation .. that is, a single "session" over which you can open multiple subchannels.

**Comment by piegamesde** [comment]
> This is one of the use-cases for Dilation

Is it? Because as far as I know Dilation is only useful within a session, but this feature request is about sending a file to multiple different persons.

**Comment by meejah** [comment]
> Is it? Because as far as I know Dilation is only useful within a session, but this feature request is about sending a file to multiple different persons.

I was replying to the comment on March 4, which sounds like it could be about 1:1 session -- although I do see that the overall feature-request is about 1:N sessions, approximately.

**Comment by zackees** [comment]
> Hi, I am thinking if it's possible for the sender to restart sending when a transaction is done. For example, when you want to deliver a file on sender A to multiple computer B1,B2,B3....Bn
> 
> Currently, the magic-wormhole exits when one transaction is done. I have to restart the sender n times and type n different keys on the receivers. Is it possible to implemented an option like `--reuse times` to keep the sender sending when a transaction is done. So that I can type the same receive code on the n servers.

+1 to this. If security is a concern, what are some thoughts on --timeout

The use case I imagine is coordination for tooling and updating server replicas. Would be quite fast!

**Comment by meejah** [comment]
@zackees re-using the _same code_ multiple times is a security issue.

However, we are working on (https://github.com/magic-wormhole/magic-wormhole-protocols/pull/23) a newer transfer protocol that can send (and/or receive) multiple distinct offers within one "session" -- that is, still a single code to set up the session but the participants can keep using the established connection for multiple transfers.

A similar and related new feature we've discussed is called "Seeds", which allows you to "re-open" a connection that you previously established with a code, but without using a code. Both sides would have to agree to remember the right information (the mailbox, and a secret) and both sides would have to try to re-establish the connection at around the same time (i.e. they still both need to be online) so you don't _completely_ eliminate the out-of-band co-ordination required, but you do get rid of the need to communicate a new code.

I don't think any of these _quite_ match the described scenario of "1:N" sending .. however, I believe they could be composed to provide a better user-experience for some cases like this (e.g. headless servers could attempt to re-establish their Seeds on some schedule, so a single "update" app could push out the next update to multiple such servers when those connections are re-established).
